### PR TITLE
Fixes for #373 and beyond

### LINKED ISSUE #373: No heartbeat and restart on MySQL 5.1.54
MySQL 5.1.54，JDK 1.8

```
bin/maxwell --host='192.168.1.106' --user='xxx' --password='xxx' --producer=stdout --include_dbs='web'
```

Replicator restart again and again:

```
15:02:30,218 INFO  TransportImpl - connecting to host: 192.168.1.106, port: 3306
15:02:30,346 INFO  TransportImpl - connected to host: 192.168.1.106, port: 3306, context: AbstractTransport.Context[threadId=19,scramble=wk4aiooauLZ^J7r[PasB,protocolVersion=10,serverHost=192.168.1.106,serverPort=3306,serverStatus=2,serverCollation=33,serverVersion=5.1.54-log,serverCapabilities=63487]
15:02:30,347 INFO  AuthenticatorImpl - start to login, user: webcast, host: 192.168.1.106, port: 3306
15:02:30,352 INFO  AuthenticatorImpl - login successfully, user: webcast, detail: OKPacket[packetMarker=0,affectedRows=0,insertId=0,serverStatus=2,warningCount=0,message=<null>]
15:02:31,108 DEBUG MysqlPositionStore - Writing binlog position to maxwell.positions: BinlogPosition[mysql-bin.000045:515930]
15:02:33,194 WARN  MaxwellReplicator - no heartbeat heard from server in 2042ms.  restarting replication.
15:02:33,195 INFO  TransportImpl - disconnected from 192.168.1.106:3306
15:02:33,196 INFO  TransportImpl - connecting to host: 192.168.1.106, port: 3306
15:02:33,214 INFO  TransportImpl - connected to host: 192.168.1.106, port: 3306, context: AbstractTransport.Context[threadId=20,scramble=Z7A$,R9eq(r&dEnw/"Z#,protocolVersion=10,serverHost=192.168.1.106,serverPort=3306,serverStatus=2,serverCollation=33,serverVersion=5.1.54-log,serverCapabilities=63487]
15:02:33,214 INFO  AuthenticatorImpl - start to login, user: webcast, host: 192.168.1.106, port: 3306
15:02:33,419 INFO  AuthenticatorImpl - login successfully, user: webcast, detail: OKPacket[packetMarker=0,affectedRows=0,insertId=0,serverStatus=2,warningCount=0,message=<null>]
15:02:33,519 WARN  MaxwellReplicator - Started replication stream outside of transaction.  This shouldn't normally happen.
15:02:35,578 WARN  MaxwellReplicator - no heartbeat heard from server in 2059ms.  restarting replication.
15:02:35,579 INFO  TransportImpl - disconnected from 192.168.1.106:3306
15:02:35,580 INFO  TransportImpl - connecting to host: 192.168.1.106, port: 3306
15:02:35,665 INFO  TransportImpl - connected to host: 192.168.1.106, port: 3306, context: AbstractTransport.Context[threadId=21,scramble=u6JQ8er04T,|OU<k'}bR,protocolVersion=10,serverHost=192.168.1.106,serverPort=3306,serverStatus=2,serverCollation=33,serverVersion=5.1.54-log,serverCapabilities=63487]
15:02:35,665 INFO  AuthenticatorImpl - start to login, user: webcast, host: 192.168.1.106, port: 3306
15:02:35,667 INFO  AuthenticatorImpl - login successfully, user: webcast, detail: OKPacket[packetMarker=0,affectedRows=0,insertId=0,serverStatus=2,warningCount=0,message=<null>]
15:02:36,160 DEBUG MysqlPositionStore - Writing binlog position to maxwell.positions: BinlogPosition[mysql-bin.000045:516212]
15:02:38,226 WARN  MaxwellReplicator - no heartbeat heard from server in 2056ms.  restarting replication.
15:02:38,227 INFO  TransportImpl - disconnected from 192.168.1.106:3306
```

### PR DESCRIPTION
**Pull request by osheroff**
Primarily this fixes #373 by disabling heartbeating on 5.1, but I also found and fixed a separate issue, where we were dumbly rewinding the replicator position after the call to `getTransactionRows` had pushed it forward; see the two separate commits in this PR if you want to see the issues separately.

@zendesk/rules

### CODE REVIEW

**Comment by lileeyao** [comment]
lgtm 👍

File: src/main/java/com/zendesk/maxwell/MaxwellContext.java [hunk 1]
```diff
@@ -31,6 +32,9 @@
 	private BinlogPosition initialPosition;
 	private CaseSensitivity caseSensitivity;
 
+	private Integer mysqlMajorVersion;
+	private Integer mysqlMinorVersion;
```
Comment (lileeyao) [inline src/main/java/com/zendesk/maxwell/MaxwellContext.java]: Why `Integer` ?
the official API says `int`

Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellContext.java]: `Integer` is nullable

File: src/main/java/com/zendesk/maxwell/MaxwellReplicator.java [hunk 2]
```diff
@@ -294,19 +298,22 @@ public RowMap getRow() throws Exception {
 					break;
 				case MySQLConstants.TABLE_MAP_EVENT:
 					tableCache.processEvent(getSchema(), this.filter, (TableMapEvent) v4Event);
+					setReplicatorPosition((AbstractBinlogEventV4) v4Event);
 					break;
 				case MySQLConstants.QUERY_EVENT:
 					QueryEvent qe = (QueryEvent) v4Event;
 					if (qe.getSql().toString().equals("BEGIN"))
 						rowBuffer = getTransactionRows();
```
Comment (dadah89) [inline src/main/java/com/zendesk/maxwell/MaxwellReplicator.java]: looks like there are some `case` conditions in `getTransactionRows` that also need to save the replicator position? (TABLE_MAP_EVENT, QUERY_EVENT, XID_EVENT)

File: src/main/java/com/zendesk/maxwell/MaxwellContext.java [hunk 3]
```diff
@@ -146,6 +150,21 @@ public Long getServerID() throws SQLException {
...
+			try ( Connection c = getReplicationConnection() ) {
+				DatabaseMetaData meta = c.getMetaData();
+				mysqlMajorVersion = meta.getDatabaseMajorVersion();
+				mysqlMinorVersion = meta.getDatabaseMinorVersion();
+			}
+		}
+	}
+
+	public boolean shouldHeartbeat() throws SQLException {
+		fetchMysqlVersion();
```
Comment (dadah89) [inline src/main/java/com/zendesk/maxwell/MaxwellContext.java]: does it matter/make sense to memoize this result?

File: src/main/java/com/zendesk/maxwell/MaxwellReplicator.java (hunk 2 above)
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellReplicator.java]: yes, that's right.  I'll make it always save the replicator position.

File: src/main/java/com/zendesk/maxwell/MaxwellContext.java (hunk 3 above)
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellContext.java]: it is already, see the `null` check on line 154.  inlining should taking care of the rest...

Comment (dadah89) [inline src/main/java/com/zendesk/maxwell/MaxwellContext.java]: ugh don't know how I missed that

File: src/main/java/com/zendesk/maxwell/MaxwellReplicator.java (hunk 2 above)
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellReplicator.java]: made `getTransactionRows` set the replicator position more

File: src/main/java/com/zendesk/maxwell/MaxwellReplicator.java [hunk 4]
```diff
@@ -202,6 +206,8 @@ private RowMapBuffer getTransactionRows() throws Exception {
 				continue;
 			}
 
+			setReplicatorPosition((AbstractRowEvent) v4Event);
```
Comment (dadah89) [inline src/main/java/com/zendesk/maxwell/MaxwellReplicator.java]: is it ok to be before the switch block? (noticed there are some return statements within it)

Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellReplicator.java]: i think so.  as soon as we grab an event from the queue we're guaranteed to process it, so if replication stalls we should always restart after it, whether we restart here or in `getRows`

**Comment by dadah89** [comment]
👍 on green

**Pull request merged**